trax/models/research/funnel_transformer.py

```diff
@@ -0,0 +1,269 @@
+# coding=utf-8
+# Copyright 2020 The Trax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Funnel Transformer model.
+
+Funnel-Transformer: Filtering out Sequential Redundancy for Efficient
+Language Processing https://arxiv.org/abs/2006.03236 """
+from trax import layers as tl
+from trax.layers.assert_shape import assert_shape
+from trax.models.transformer import _EncoderBlock, _FeedForwardBlock
+
+
+def _InternalMaxPool(arr):
+  shape = arr.shape
+  arr = arr.reshape((*shape[:-1], int(shape[-1] / 2), 2))
+  arr = arr.max(axis=-1, keepdims=False)
+  return arr
+
+
+@assert_shape('bld->bSd')
+def PoolLayer(pool_layer=tl.AvgPool,
+              pool_size=(2,),
+              strides=(2,),
+              separate_cls=True):
+  return tl.Serial(
+      tl.Branch(
+          tl.Fn('select_cls_token', lambda x: x[:, :1, :]),
+          tl.Serial(
+              tl.Fn('rest_tokens', lambda x: x[:, 1:, :]),
+              pool_layer(pool_size, strides)
+          )
+      ),
+      tl.Concatenate(axis=1)
+  ) if separate_cls else pool_layer(pool_size, strides)
```

Comment:
I would prefer to see here if-else statement rather than conditional expression; this Serial is quite long anyway. Can you split it?

Reply:
done